File: Algorithms/src/ConsoleTester/Day01.fs
module Day01

open Common
open System

[<Literal>]
let fileName = "..\\..\\..\\..\\..\\day1.txt"

///
/// Day 1 of: https://adventofcode.com/2024
///
let Run =

    printfn "\nDay01 of advent of code 2024\n"

    let data = FileReader.lines fileName

    let left =
        data
        |> Seq.map (fun line -> splitIntoArrayOfIntegers line " " 0)
        |> Seq.toList
        |> List.sort

    let right =
        data
        |> Seq.map (fun line -> splitIntoArrayOfIntegers line " " 1)
        |> Seq.toList
        |> List.sort

    let distance2 =
        List.zip left right |> List.map (fun (a, b) -> abs (a - b)) |> List.sum

    let distance =
        List.map2 (fun (left: int) (right: int) -> Math.Abs(left - right)) left right
        |> List.sum

    let similarity = List.filter (fun x -> List.contains x left) right |> List.sum

    printfn "Distance: %i" distance

    printfn "Distance: %i" distance2

    printfn "Similarity: %i" similarity

File: Algorithms/src/Algorithms/Day01.cs
namespace Algorithms
{
    internal static class Day01
    {
        /// <summary>
        /// Example/solution for advent-of-code-2024
        /// Day One
        /// </summary>
        internal static void Run(bool testing = false)
        {
            (List<int> left, List<int> right) = testing ? LoadTestData() : LoadLists("..\\..\\..\\..\\..\\day1.txt");

            WriteLine(new string('*', 80));

            var totalDistance = left.Order()
                /*  Applies a specified function to the corresponding elements of
                 *  two sequences, producing a sequence of results
                 */
                .Zip(right.Order(), (x, y) =>
                {
                    WriteLine($"Left: {x}\tRight: {y}");
                    return Math.Abs(x - y);
                }
                )
                .Sum();

            WriteLine(new string('*', 80));

            /*
             * For every element in the left
             * find all corresponding elements on the right (if any)
             * and sum their values
             * or count the number of occurrences and multiply the count
             * by the value in the left-hand-side
             * (same thing), in effect
             */
            int similarityScore = left
                /*  Correlates the elements of two sequences based on matching keys.
                 *  The default equality comparer is used to compare keys.
                 */
                .Join(right, l => l, r => r, (l, r) =>
                {
                    WriteLine($"Left: {l}\tRight: {r}");
                    return r;
                })
                .Sum();

            WriteLine($"Total Items: {left.Count}");
            WriteLine($"Total Distance: {totalDistance}");
            WriteLine($"Similarity Scores: {similarityScore}");
        }

        private static (List<int> a, List<int> b) LoadLists(string fileName) =>
            CommonFuncs.LoadFileData(fileName).Select(CommonFuncs.ParseIntsNoSign).Transpose().ToPair();

        private static (List<int> left, List<int> right) LoadTestData() =>
            ([2, 2, 3, 3, 3, 4, 7, 9], [2, 3, 3, 3, 3]);

    }
}
